# Data Consistency Check

In [2]:
#Importing libraries

import pandas as pd
import numpy as np
import os

In [3]:
path = r'C:\Users\Güldeniz\Desktop\Career Foundry\2. Data Immersion\Achievement 4\Instacart  Basket Analysis'

In [4]:
#products data frame
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
#orders data frame
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [8]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

### Mixed-Type Data

In [9]:
# Create a datafrane
df_test = pd.DataFrame()

In [10]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [11]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [12]:
#Check for mixed types (always check for mixed-type columsn before moving forward)
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply (type)).any(axis = 1)
    if len (df_test[weird])>0:
        print (col)

mix


In [13]:
#Changing from numeric to string
df_test['mix']= df_test ['mix'].astype('str')

### Missing Values

In [5]:
#Finding missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [6]:
#To actually view these 16 values, you can create a subset of the dataframe containing only the values in question
df_nan = df_prods[df_prods['product_name'].isnull()== True]

In [21]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [22]:
# Compare the number of rows in your current dataframe with the number in your subset once the missing rows have been removed:

df_prods.shape

(49693, 5)

In [7]:
#Subset without the missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull()== False]

In [24]:
df_prods_clean.shape

(49677, 5)

In [ ]:
# Dropping the  missing values: df_prods.dropna(inplace=True)
# Drop only the NaNs from a particular column: df_prods.dropna(subset = ['product_name'], inplace = True])
# inplace = True overwrites the original dataset (risky), inplace=False will only return a 'view' of the changed 
# ...dataframe, original dataframe left untouched.

### Duplicates

In [26]:
df_prods.head(10)

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   
5           6                                       Dry Nose Oil        11   
6           7                     Pure Coconut Water With Orange        98   
7           8                  Cut Russet Potatoes Steam N' Mash       116   
8           9                  Light Strawberry Blueberry Yogurt       120   
9          10     Sparkling Orange Juice & Prickly Pear Beverage       115   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3  
5             11     2.6  
6              7     4.4  
7              1     1.1  
8             16     7.0  
9              7     8.4

In [8]:
# The duplicated() function is what identifies duplicate rows.
df_dups = df_prods_clean [df_prods_clean.duplicated()]

In [29]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [30]:
# Addressing the duplicates: df.drop_duplicates()
#check the current number of rows in df_prods_clean dataframe by shape function.
df_prods_clean.shape

(49677, 5)

In [9]:
#Creating a dataframe that does not include the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [33]:
df_prods_clean_no_dups.shape  # 5 duplicate values are dropped.

(49672, 5)

### Tidying Up and Exporting Changes

In [10]:
df_prods_cleaned = df_prods_clean_no_dups

In [11]:
#exporting cleaned products.csv dataframe
df_prods_cleaned.to_csv(os.path.join(path,'02 Data','Prepared Data','products_cleaned.csv'), index=False)

# Task 4.5 Data Consistency Check

### 2. Run df.describe function on your df_prods dataframe

In [12]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [37]:
df_prods_cleaned.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

##### => Cheking the count, the dropped values show themselves. There is no change, however, in min and max values. Dataframe looks OK!

### 3. Check for mixed-type data in df_ords dataframe.

In [47]:
df_ords.head(50)

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0    2539329        1    prior             1          2                  8   
1    2398795        1    prior             2          3                  7   
2     473747        1    prior             3          3                 12   
3    2254736        1    prior             4          4                  7   
4     431534        1    prior             5          4                 15   
5    3367565        1    prior             6          2                  7   
6     550135        1    prior             7          1                  9   
7    3108588        1    prior             8          1                 14   
8    2295261        1    prior             9          1                 16   
9    2550362        1    prior            10          4                  8   
10   1187899        1    train            11          4                  8   
11   2168274        2    prior             1          2                 11   
12   1501582        2    prior             2          5                 10   
13   1901567        2    prior             3          1                 10   
14    738281        2    prior             4          2                 10   
15   1673511        2    prior             5          3                 11   
16   1199898        2    prior             6          2                  9   
17   3194192        2    prior             7          2                 12   
18    788338        2    prior             8          1                 15   
19   1718559        2    prior             9          2                  9   
20   1447487        2    prior            10          1                 11   
21   1402090        2    prior            11          1                 10   
22   3186735        2    prior            12          1                  9   
23   3268552        2    prior            13          4                 11   
24    839880        2    prior            14          3                 10   
25   1492625        2    train            15          1                 11   
26   1374495        3    prior             1          1                 14   
27    444309        3    prior             2          3                 19   
28   3002854        3    prior             3          3                 16   
29   2037211        3    prior             4          2                 18   
30   2710558        3    prior             5          0                 17   
31   1972919        3    prior             6          0                 16   
32   1839752        3    prior             7          0                 15   
33   3225766        3    prior             8          0                 17   
34   3160850        3    prior             9          0                 16   
35    676467        3    prior            10          3                 16   
36    521107        3    prior            11          0                 18   
37   1402502        3    prior            12          1                 15   
38   2774568        3     test            13          5                 15   
39   3343014        4    prior             1          6                 11   
40   2030307        4    prior             2          4                 11   
41    691089        4    prior             3          4                 15   
42     94891        4    prior             4          5                 13   
43   2557754        4    prior             5          5                 13   
44    329954        4     test             6          3                 12   
45   2717275        5    prior             1          3                 12   
46   1909121        5    prior             2          0                 16   
47   2267326        5    prior             3          3                 18   
48    157374        5    prior             4          1                 18   
49   2196797        5    train             5          0                 11   

    days_since_prior_orde

In [42]:
df_ords.tail(50)

order_id  user_id eval_set  order_number  order_dow  \
3421033   1949208   206208    prior            15          2   
3421034   3049681   206208    prior            16          2   
3421035     11859   206208    prior            17          1   
3421036    447879   206208    prior            18          3   
3421037   2467456   206208    prior            19          4   
3421038    842771   206208    prior            20          0   
3421039   2471606   206208    prior            21          5   
3421040   2922221   206208    prior            22          1   
3421041   3128968   206208    prior            23          5   
3421042   2573371   206208    prior            24          5   
3421043   2804556   206208    prior            25          2   
3421044   2079142   206208    prior            26          2   
3421045    183408   206208    prior            27          2   
3421046   3038194   206208    prior            28          0   
3421047   3412250   206208    prior            29          4   
3421048    726186   206208    prior            30          0   
3421049   2523609   206208    prior            31          4   
3421050    756221   206208    prior            32          6   
3421051   2254244   206208    prior            33          6   
3421052   2017995   206208    prior            34          6   
3421053   2789700   206208    prior            35          3   
3421054    844592   206208    prior            36          6   
3421055   1541132   206208    prior            37          2   
3421056   2808240   206208    prior            38          0   
3421057   3027766   206208    prior            39          2   
3421058   3356245   206208    prior            40          5   
3421059    442304   206208    prior            41          2   
3421060   2675140   206208    prior            42          1   
3421061    167903   206208    prior            43          4   
3421062   2393201   206208    prior            44          6   
3421063   3292671   206208    prior            45          2   
3421064   3059777   206208    prior            46          1   
3421065   2239861   206208    prior            47          3   
3421066   1285346   206208    prior            48          1   
3421067   1882108   206208    prior            49          1   
3421068    803273   206208     test            50          5   
3421069   3154581   206209    prior             1          3   
3421070   1889163   206209    prior             2          3   
3421071   1542354   206209    prior             3          5   
3421072    688306   206209    prior             4          1   
3421073   2307371   206209    prior             5          4   
3421074   3186442   206209    prior             6          0   
3421075    550836   206209    prior             7          2   
3421076   2129269   206209    prior             8          3   
3421077   2558525   206209    prior             9          4   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  
3421033                 14                     6.0  
3421034                 15                     7.0  
3421035                 13                     6.0  
3421036                 23                     9.0  
3421037                 13                    15.0  
3421038                  8                    10.0  
3421039                  9                     5.0  
3421040                 12                     3.0  
3421041                 16                     4.0  
3421042                  9                     7.0  
3421043                 15                    11.0  
3421044                 16                     0.0  
3421045                 14                     7.0  
34

### 4. If you find mixed-type data, fix it.
##### => Checking the df-ords dataframe, I could not see any mixed-type data.

### 5.Run a check for missing values in your df_ords dataframe

In [6]:
#Finding missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#####  => The only column with missing values is days_since_prior_order. There are 206209 missing values in this column, which is a lot! The reason for this many mssing values can be the fact that, many of the orders can be the first orderof the user. That means, there is no value to give for numbers of days since the prior order. 

### 6. Address the missing values using an appropiate method.
##### => For the missing values in 'days_since_prior_order' I would chose to impute the missing values (NaN) with value 0. Since these missing values should represent new customers, without prior orders. 

In [7]:
#checking the shape of the dataframe

df_ords.shape

(3421083, 7)

In [8]:
#Imputation with a 0 value:

df_ords['days_since_prior_order']=df_ords['days_since_prior_order'].fillna(0)

In [9]:
#Controlling the imputation

df_ords.isnull().sum()

order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

### 7. Run a check for duplicate values in your df_ords data

In [10]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [11]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

##### => there are no duplicate values in orders dataframe. 

### 8. Address the duplicates using an appropriate method
##### => There are no duplicate values. 
##### => If there were any duplicate values, I would  use the function: df_ords_no_dups = df_ords.drop_duplicates().  I would also check the row and column numbers before and after I run the dropping duplicates function.

### 9. Export your final cleaned df_prods and df_ords data as "".csv" files.

In [12]:
#Exporting df_ords dataframe:

df_ords_cleaned = df_ords
df_ords_cleaned.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_cleaned.csv'), index=False)

##### => 'products_clean.csv' file was exported in earlier stages (see. Tidying up and exporting changes)